include/deal.II/fe/fe_p1nc.h

```diff
@@ -0,0 +1,316 @@
+ *
+ * \phi(m_0) + \phi(m_1) = \phi(m_2) + \phi(m_3).
+ *
+ * Conversely if 4 values at midpoints satisfying the dice rule are just given,
+ * then there always exists the unique linear function which coincides with 4 midpoints values.
+ *
+ * Due to the dice rule, three values at any three midpoints can determine the last value at the last midpoint.
+ * It means that the genuine number of (independent) DOFs on a quad is 3,
+ * and it is the same number to the dimension of the linear polynomial space in 2D.
+
+
+ * <h3>Shape functions</h3>
+
+ *  @verbatim
+ *  2---------|---------3
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  -                   -
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  0---------|---------1
+ *  @endverbatim
+
+ * For each vertex v_j of given quad, there are two edges of which v_j is one of end points.
+ * Consider a linear function such that 0.5 value at two midpoints of such edges,
+ * and 0.0 at two midpoints of other edges.
+ * Note that the set of these values satisfies the dice rule which is described above.
+ * We denote such a function assoicated with vertex v_j by \phi_j.
+
+ * The canonical (local) basis functions are given as any three shape functions of
+ * the following four linear functions:
+
+ * <ul>
+ * <li> shape function \phi_0:
+ *  @verbatim
+ *  +--------0.0--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.5                 0.0
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.5--------+
+ *  @endverbatim
+
+ * <li> shape function \phi_1:
+ *  @verbatim
+ *  +--------0.0--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.0                 0.5
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.5--------+
+ *  @endverbatim
+
+ * <li> shape function \phi_2:
+ *  @verbatim
+ *  +--------0.5--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.5                 0.0
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+
+ * <li> shape function \phi_3:
+ *  @verbatim
+ *  +--------0.5--------+
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ * 0.0                 0.5
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  |                   |
+ *  +--------0.0--------+
+ *  @endverbatim
+
+ * </ul>
+
+ * Note that above shape functions are constructed on each cell, not on the reference cell only.
```

Comment:
To construct shape functions on each cell,
this element does not use "pull-back" from the reference cell.

On usual finite elements, shape values on the reference cell are computed at first,
and shape values on each cell are computed based on associated values on the reference cell.

But this element computes shape values on each cell independently.
Shape values on one cell are not related with those on other cells at all.